### PR TITLE
Issue 1951 Check that generator expressions are safe

### PR DESCRIPTION
**Pull request by GibranHL0**
# I have made things!

## Checklist

- [x] I have double checked that there are no unrelated changes in this pull request (old patches, accidental config files, etc)
- [x] I have created at least one test case for the changes I have made
- [x] I have updated the documentation for the changes I have made
- [x] I have added my changes to the `CHANGELOG.md`

## Related issues

🙏 Please, if you or your company is finding wemake-python-styleguide valuable, help us sustain the project by sponsoring it transparently on https://opencollective.com/wemake-python-styleguide. As a thank you, your profile/company logo will be added to our main README which receives hundreds of unique visitors per day.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -109,6 +109,7 @@ Semantic versioning in our case means:
 - Forbids inconsistent structuring of multiline comprehensions
 - Forbids to use unpythonic getters and setters such as `get_attribute` or `set_attribute`
 - Now `credits`, `license`, and `copyright` builtins are free to shadow
+- Forbids the reassignment of variables that affect the generator expression
```
Comment (sobolevn) [inline CHANGELOG.md:112]: This would be released under `0.16.0`, not `0.15.0`

File: wemake_python_styleguide/violations/consistency.py [hunk 2]
```diff
@@ -2343,3 +2345,37 @@ class AssignToSliceViolation(ASTViolation):
...
+
+    Solution:
+        Either do not reassign new values to the variables that affect the
+        generator expression or implement it before reassigning.
+
+    Example:
+        # Correct
+        a = 1
+        b = (a * i for i in range(5))
+        c = (sum(b))
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/consistency.py:2368]:
```suggestion
        c = sum(b)
```

File: wemake_python_styleguide/violations/consistency.py [hunk 3]
```diff
@@ -2343,3 +2345,37 @@ class AssignToSliceViolation(ASTViolation):
...
+        Reassigning values to variables that affect a generator expression
+        before the generator expression is used will produce a different
+        result from what it is expected.
+
+    Solution:
+        Either do not reassign new values to the variables that affect the
+        generator expression or implement it before reassigning.
+
+    Example:
+        # Correct
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/consistency.py:2365]: Can you please use more meaningful names here? We try to write our correct examples as close to the real-world code as possible.

File: tests/test_visitors/test_ast/test_loops/test_comprehensions/test_unsafe_generator_expression.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    UnsafeGeneratorExpressionViolation,
+)
+from wemake_python_styleguide.visitors.ast.statements import (
+    UnsafeGeneratorExpressionVisitor,
+)
+
+unsafe_function = """
+first_value = 1
+expression = (first_value * index for index in range(5))
+first_value = 2
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_loops/test_comprehensions/test_unsafe_generator_expression.py:13]: What about changing `index`? Let's test that as well.

File: wemake_python_styleguide/violations/consistency.py [hunk 5]
```diff
@@ -2343,3 +2345,37 @@ class AssignToSliceViolation(ASTViolation):
...
+        b = (a * i for i in range(5))
+        c = (sum(b))
+
+        # Incorrect
+        a = 1
+        b = (a * i for i in range(5))
+        a = 2
+        c = (sum(c))
+
+    .. versionadded:: 0.15.0
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/consistency.py:2376]: 0.16.0

File: wemake_python_styleguide/visitors/ast/statements.py [hunk 6]
```diff
@@ -498,3 +508,108 @@ def _check_tuple_arguments_types(
...
+
+
+@final
+class UnsafeGeneratorExpressionVisitor(BaseNodeVisitor):
+    """Checks if a generator expression is safe."""
+
+    def __init__(self, *args, **kwargs) -> None:
+        """Creates the lists that would hold the variables."""
+        super().__init__(*args, **kwargs)
+        self.inside_variables: List[str] = []
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/statements.py:520]: We use `_inside_variables` (protected notation) for visitor attributes.

File: wemake_python_styleguide/visitors/ast/statements.py [hunk 7]
```diff
@@ -498,3 +508,108 @@ def _check_tuple_arguments_types(
...
+        self.created_variables: List[str] = []
+        self.affecting_variables: List[str] = []
+
+    def visit(self, node: ast.AST) -> None:
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/statements.py:524]: Why don't you use `visit_any_assign` here? It does the same thing.

File: wemake_python_styleguide/visitors/ast/statements.py [hunk 8]
```diff
@@ -498,3 +508,108 @@ def _check_tuple_arguments_types(
...
+        """Checks if the generator expression has affecting variables."""
+        assignments = [
+            sub_node
+            for sub_node in ast.walk(node)
+            if isinstance(sub_node, ast.Assign)
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/statements.py:529]: You only check for `Assign`, what about `AugAssign` like `+= 1` or assigning expression like `(x := 1)`? We need to support them as well.

File: wemake_python_styleguide/visitors/ast/statements.py [hunk 9]
```diff
@@ -498,3 +508,108 @@ def _check_tuple_arguments_types(
...
+                created = self.created_variables
+
+                self.affecting_variables = ([
+                    name
+                    for name in inside
+                    if name not in inside or name not in created
+                ])
+
+                if self.affecting_variables:
+                    self._check_gen_usage(node, assignments)
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/statements.py:552]: You are missing `self.generic_visit(node)` call at the bottom. It is required.

File: wemake_python_styleguide/visitors/ast/statements.py [hunk 10]
```diff
@@ -498,3 +508,108 @@ def _check_tuple_arguments_types(
...
+
+    def _get_inside_variables(self, node: ast.GeneratorExp) -> None:
+        sub_node = node.elt
+        self.inside_variables = [
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/statements.py:556]: We already have this logic as a function: https://github.com/wemake-services/wemake-python-styleguide/blob/master/wemake_python_styleguide/logic/walk.py#L56

File: wemake_python_styleguide/visitors/ast/statements.py [hunk 11]
```diff
@@ -498,3 +508,108 @@ def _check_tuple_arguments_types(
...
+        for generator in assignments:
+            if isinstance(generator.value, ast.GeneratorExp):
+                for target in generator.targets:
+                    if isinstance(target, ast.Name):
+                        gen_target_id_lineno = self._get_id_lineno(target)
+                        gen_positions.append(gen_target_id_lineno)
+
+        return gen_positions
+
+    def _get_id_lineno(self, node: ast.Name):
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/statements.py:589]: Missing return type

File: wemake_python_styleguide/visitors/ast/statements.py [hunk 12]
```diff
@@ -498,3 +508,108 @@ def _check_tuple_arguments_types(
...
+        return (node.id, node.lineno)
+
+    def _check_gen_usage(self, node: ast.AST, assignments: List[ast.Assign]):
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/statements.py:592]: Missing return type

**Review by sobolevn — CHANGES_REQUESTED** [review]
Great work! Thanks! 💪 

It needs a bit of polishing and I would be happy to merge it.

**Comment by GibranHL0** [comment]
Sure, I'll be working on them today, and I hope later today or tomorrow early commit the requested changes. 🤓

File: tests/test_visitors/test_ast/test_loops/test_comprehensions/test_unsafe_generator_expression.py (hunk 4 above)
Comment (GibranHL0) [reply tests/test_visitors/test_ast/test_loops/test_comprehensions/test_unsafe_generator_expression.py:13]: I believe that the index cannot be changed as the variable is load within the generator expression. Even though, I'll include a test

File: wemake_python_styleguide/visitors/ast/statements.py (hunk 7 above)
Comment (GibranHL0) [reply wemake_python_styleguide/visitors/ast/statements.py:524]: I didn't consider it 🤔, I'm changing it!

File: wemake_python_styleguide/violations/consistency.py [hunk 13]
```diff
@@ -2363,17 +2363,17 @@ class UnsafeGeneratorExpressionViolation(ASTViolation):
 
     Example:
         # Correct
-        a = 1
-        b = (a * i for i in range(5))
-        c = (sum(b))
+        suffix = 'ish'
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/consistency.py:2366]: Great one! 👍

File: wemake_python_styleguide/visitors/ast/statements.py (hunk 8 above)
Comment (GibranHL0) [reply wemake_python_styleguide/visitors/ast/statements.py:529]: Sure, it needs to have AugAssign support, but it'll add too much complexity with the current Visitor.

In the case of the assigning expression, there is a Violation that prevents the usage of `(x := 1)` , WPS332 Found Walrus operator

**Comment by GibranHL0** [comment]
I uploaded a new version of the UnsafeGeneratorExpressionVisitor that fixes the issues with complexity and style. But I'd like to add a new visitor that adds support for the AugAssign. I'll upload it in less than a day! 👍

**Comment by sobolevn** [comment]
> But I'd like to add a new visitor that adds support for the AugAssign

This sounds like not the best idea. 
The better way: move logic to `logic/` and use  a single visitor with all the types you need to call it.

**Comment by GibranHL0** [comment]
I figured out a new way to support `AugAssign` within the visitor and not creating a logic method outside. I also included new tests.(/7-7)/

File: tests/test_visitors/test_ast/test_loops/test_comprehensions/test_unsafe_generator_expression.py [hunk 14]
```diff
@@ -62,11 +62,28 @@
 exponent = 10
 """
 
+unsafe_aug_value = """
+values = [1, 2, 3, 4, 5]
+exponent = 5
+power_expr = (value * exponent for value in values)
+exponent += 1
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_loops/test_comprehensions/test_unsafe_generator_expression.py:69]: What about changing `values` here? Like `values.append(1)`
Should we allow it?

**Review by sobolevn — COMMENTED** [review]
Great progress! 👍

File: tests/test_visitors/test_ast/test_loops/test_comprehensions/test_unsafe_generator_expression.py (hunk 14 above)
Comment (GibranHL0) [reply tests/test_visitors/test_ast/test_loops/test_comprehensions/test_unsafe_generator_expression.py:69]: I think we should not because it affects the expected behavior of the expression. Can I create a new rule request to prevent it? 🤔

Comment (sobolevn) [reply tests/test_visitors/test_ast/test_loops/test_comprehensions/test_unsafe_generator_expression.py:69]: I am not sure that it is "expected" because some people might be getting an error here. Let me give you an example:

```python
>>> values = [1, 2, 3]
>>> listcomp = [a * 2 for a in values]
>>> values.append(0)
>>> values, listcomp
([1, 2, 3, 0], [2, 4, 6])
```

Here's how it works with lists: no unexpected problems.
Now, the same with `gen`:

```python
>>> values = [1, 2, 3]
>>> gen = (a * 2 for a in values)
>>> values.append(0)
>>> values, list(gen)
([1, 2, 3, 0], [2, 4, 6, 0])
```

Now, we can see that changing `values` changes how `gen` works. It is not consistent with list comprehensions. If a persons just refactored list comprehension to be a generator expression with a code like this - there would be an error.

And we can catch it!

I recommend to try to visit `ast.Name` with usage / reassining context. It might be easier that our current appoach.

Comment (GibranHL0) [reply tests/test_visitors/test_ast/test_loops/test_comprehensions/test_unsafe_generator_expression.py:69]: Yes, I think that we should prevent it!

But, with the current visitor, it would be difficult to implement without making it complex because the `ast.Name` of values have `ctx` `ast.Load` in comparisons, functions or creating other list comprehensions and gens that not affect the `gen`.
```python
>>> values = [1, 2, 3, 4, 5]
>>> gen = (a * 2 for a in values)
>>> if 2 in values:
>>>    addition = sum(values)
>>> values, list(gen)
([1, 2, 3], [2, 4, 6]
```
Then we should check that the `ast.Attribute` append of `gen` is not being used within assignment and usage. That would need at least 2 new methods `visit_Attribute` and `_check_append` but the current visitor is at maximum allowed method capacity.

Also, I believe that a new violation would make the error easier to understand as the current violation only explains the variables that affect a generator expression. The `UnsafeAppendUsageViolation` would point out the forbidden usage of append within assignment and call for lists that affect a generator expression.

Comment (sobolevn) [reply tests/test_visitors/test_ast/test_loops/test_comprehensions/test_unsafe_generator_expression.py:69]: Will the current implementation work for:

```python
self.x = 1
gen = (self.x * 2 for a in values)
self.x = 2
...
```

?

> That would need at least 2 new methods visit_Attribute and _check_append but the current visitor is at maximum allowed method capacity.

Don't worry about it for now, let's concentrate on making this work for all general and edge cases first, then we can refactor the working solution into a simplier version.

Comment (GibranHL0) [reply tests/test_visitors/test_ast/test_loops/test_comprehensions/test_unsafe_generator_expression.py:69]: Sure then, I'll be working on append support. 📝

On the other hand, the current implementation doesn't work with `self` variables. So, I'll update it as well with new tests 🤠

**Comment by sobolevn** [comment]
Thanks for the PR, but it now has merge conflicts.
